# Lesson 3 Variables and dataframes

These exercises are to help you check your understanding during class. They will not be collected.

## Useful links
### Class notes
Link on Brightspace or<br>
Tiny URL: https://tinyurl.com/Fall21CORE111


### Poll Everywhere
The following links are also on NYU Classes<br>
Login: https://www.polleverywhere.com/auth/saml/nyu_main<br>
Respond: https://pollev.com/clarkson<br>

## Part 1: Lists

### Exercise 0 - Access notebook and copy it
* Save your own copy of this notebook by using the menu opetion File > Save As. 
* Now, **refresh the page**. The red icon in the upper right should no longer be visible. 

### Exercise 1 - Review of lists

Last class, we learned to make lists using both c() and number : number. We also learned to do perform arithemtic with lists.

Without using Copy & Paste create a list whose entries are the following numbers:
10 15 20 25 30 35 40 45 50 55.

In [ ]:
(2:11)*5

### Exercise 2 - Ice cream sale!!!

An icecream truck sells many different types of ice cream. The prices of each type of ice cream are stored in a list called `ice_cream_prices`in the code cell below. Today the icecream is on sale! The amount that each item is discounted is stored in the list called `discounts`. The order of the discounts corresponds to the order of the prices. 

Create a new list with the sale prices of each item and store the list in a variable called `sale_prices`.

In [1]:
ice_cream_prices <- c(1.25, 3, 5.99, 1.75, 6.75, 4.99, 7.09, 3.55, 2)
discounts <- c(0.25, 1, 0.50, 0.5, 1, 0, 1, 0.25, 0)
# notice that R gives no output when you assign values to variables
# you can see that the cell has been exicuted if there is a number in the brackets: In []

In [ ]:
# if we want to check tha thte values are in a variable, we can exicute just the variable name
discounts

In [ ]:
sale_prices <- ice_cream_prices - discounts
sale_prices

### Exercises 3 - More ice cream 
Use `sale_prices` from excercise 2 and R functions to determine the ansers to the following questions:
* How many types of ice cream does the truck sell?
* What is the cost of the most exspensive frozen treat?
* What is the cost of the cheapest ice cream option?
* What is the average cost of an item from the truck rounded to the nearest cent?

In [ ]:
length(sale_prices) # number of types of ice cream
max(sale_prices) # cost of most expensive
min(sale_prices) # cheapest price
mean(sale_prices) # average cost

## Part 2: Old faithful

### Exercise 4 - Old Failthful

One of the data sets in the datasets package is called `faithful`. Use the documentation on this dataset to answer the following question. Wrtie your answer in the markdown cell below. 

Why is there reason to believe that this data includes errors introduced by humans?

In [10]:
?faithful

Warning message in seq_len(head.end.idx):
“first element used of 'length.out' argument”
ERROR while rich displaying an object: Error in seq_len(head.end.idx): argument must be coercible to non-negative integer

Traceback:
1. FUN(X[[i]], ...)
2. tryCatch(withCallingHandlers({
 .     if (!mime %in% names(repr::mime2repr)) 
 .         stop("No repr_* for mimetype ", mime, " in repr::mime2repr")
 .     rpr <- repr::mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler), error = outer_handler)
3. tryCatchList(expr, classes, parentenv, handlers)
4. tryCatchOne(expr, names, parentenv, handlers[[1L]])
5. doTryCatch(return(expr), name, parentenv, handler)
6. withCallingHandlers({
 .     if (!mime %in% names(repr::mime2repr)) 
 .         stop("No repr_* for mimetype ", mime, " in repr::mime2repr")
 .     rpr <- repr::mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_conte

### Exercise 5 - The last observation

How long was the last erruption recorded in the fialthful data set? Write you answer in the markdown cell below. Don't forget units!

In [11]:
faithful

eruptions waiting
1   3.600     79     
2   1.800     54     
3   3.333     74     
4   2.283     62     
5   4.533     85     
6   2.883     55     
7   4.700     88     
8   3.600     85     
9   1.950     51     
10  4.350     85     
11  1.833     54     
12  3.917     84     
13  4.200     78     
14  1.750     47     
15  4.700     83     
16  2.167     52     
17  1.750     62     
18  4.800     84     
19  1.600     52     
20  4.250     79     
21  1.800     51     
22  1.750     47     
23  3.450     78     
24  3.067     69     
25  4.533     74     
26  3.600     83     
27  1.967     55     
28  4.083     76     
29  3.850     78     
30  4.433     79     
⋮   ⋮         ⋮      
243 4.933     86     
244 2.900     63     
245 4.583     85     
246 3.833     82     
247 2.083     57     
248 4.367     82     
249 2.133     67     
250 4.350     74     
251 2.200     54     
252 4.450     83     
253 3.567     73     
254 4.500     73     
255 4.150     88     
256 3.817     80     
257 3.917     71     
258 4.450     83     
259 2.000     56     
260 4.283     79     
261 4.767     78     
262 4.533     84     
263 1.850     58     
264 4.250     83     
265 1.983     43     
266 2.250     60     
267 4.750     75     
268 4.117     81     
269 2.150     46     
270 4.417     90     
271 1.817     46     
272 4.467     74

The last eruption was 4.467 minutes.

In [12]:
str(faithful)

'data.frame':	272 obs. of  2 variables:
 $ eruptions: num  3.6 1.8 3.33 2.28 4.53 ...
 $ waiting  : num  79 54 74 62 85 55 88 85 51 85 ...


In [13]:
faithful$eruptions[272]

[1] 4.467

### Exercise 6 - Change the units

The faithful data set gives the duration of the erruptions in minutes and fractions of a minute. It would be easer to understad the data if we converted the decimals to seconds. Demonstrate that this could be done by finding the duration of the 13th recorded eruption in therms of minuts and seconds. 

In [14]:
faithful$eruptions[13]

[1] 4.2

In [16]:
#0.2  minutes to seconds
# 60 seconds in a minute
0.2*60

[1] 12

The 13th eruption lasted 4 minutes and 12 seconds.

## Part 3: UC Berkeley

### Exercise 7 - Load Berkeley addmissions data

Execute the command in the code cell below to load the Berkeley admisions data. Then varify that it loaded correctly by excuting a command that displays all or part of the data. 

In [19]:
berkeley <- read.csv('~/shared/data/berkeley73.csv')

### Exercise 8 - Admissons rates

Add two new columns to the berkeley data frame. One that contains the percentage of men who were admitted by each department and a second that contains the percentage of women that were admitted by each department. 

## Examples

### Example 1: Variables and list fucntions

In [ ]:
numbers <- c(5.8, 3.52, 7.81, 0.4, 4.79, 3.04, 4.89, 5.23, 2.37, 4.32) 
# generated using a random number generator
mean(numbers)

In [35]:
max(numbers)

[1] 7.81

### Example: Airquality data set

In [8]:
library(help = "datasets")

In [17]:
?airquality

In [18]:
head(airquality,3) # gives the first few rows of a data frame

Ozone Solar.R Wind Temp Month Day
1 41    190      7.4 67   5     1  
2 36    118      8.0 72   5     2  
3 12    149     12.6 74   5     3

In [14]:
mean(airquality$temp) # Note that R is sensitice to campialization

Warning message in mean.default(airquality$temp):
“argument is not numeric or logical: returning NA”

[1] NA

In [15]:
mean(airquality$Temp)

[1] 77.88235

The average high temperature in NYC from May to Septemerin 1973 was $77.9^\circ$ F.

In [18]:
max(airquality$Temp)

[1] 97

The hottest day was $97^\circ$ F.

### Example 3: Change the units

We can use the formula below to convert a temperature $f$ in degrees Fahrenheit to the equivalent temperature $c$ in degrees Celcius
$$(f − 32) \cdot \frac{5}{9} = c$$

In [19]:
f <- 97
(f-32)*5/9

[1] 36.11111

The hottest day was $36.1^\circ$ C.

In [23]:
airquality$Celcius <- (airquality$Temp - 32)*5/9
head(airquality)

Ozone Solar.R Wind Temp Month Day Celcius 
1 41    190      7.4 67   5     1   19.44444
2 36    118      8.0 72   5     2   22.22222
3 12    149     12.6 74   5     3   23.33333
4 18    313     11.5 62   5     4   16.66667
5 NA     NA     14.3 56   5     5   13.33333
6 28     NA     14.9 66   5     6   18.88889